### PR TITLE
Consider installed packages in solve

### LINKED ISSUE #88: Solver reports `RattlerUnsatisfiableError` when updating packages that are not available in the current set of channels
### Checklist

- [x] I added a descriptive title
- [x] I searched open reports and couldn't find a duplicate

### What happened?

The rattler solver will report a `RattlerUnsatisfiableError` when updating an environment that contains packages that are not available in the current list of channels. You can observe this error if you create a new environment, then remove all channels and try to update the environment. For example:

Consider the following condarc:
```
channels:
  - main
  - conda-forge
solver: rattler
```

Run the conda commands to create a new environment with python and algoliasearch. algoliasearch is a package that exists in conda-forge but not main.
```
$ conda create -n one python algoliasearch
```

Now update the condarc to remove the conda-forge channel
```
channels:
  - main
solver: rattler
```

Try to update the conda environment
```
$ conda update -n one --all                                            
Channel: N/A
Platform: linux-64
Target prefix: /home/sophia/projects/conda/devenv/Linux/x86_64/envs/devenv-3.10-miniconda/envs/one
Collecting package metadata (repodata.json): done
Solving environment: failed

RattlerUnsatisfiableError: Cannot solve the request because of: No candidates were found for algoliasearch *.
```

Running the same command with libmamba solver does not produce an error
```
$ conda update -n one --all --solver libmamba                          
Channels:
 - 
Platform: linux-64
Collecting package metadata (repodata.json): done
Solving environment: done

# All requested packages already installed.
```

### PR DESCRIPTION
**Pull request by soapy1**
### Description

This change allows packages installed in the environment to be considered in the solve by including them in them in the `RattlerIndexHelper`. We do not include virtual packages in the set of installed packages.

With this change, removing a channel from an environment and running an update no longer produces an Unsatisfiable error. 

fixes https://github.com/conda/conda-rattler-solver/issues/88

### Checklist - did you ...

- [x] Add a file to the `news` directory ([using the template](https://github.com/conda/conda-rattler-solver/blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [x] Add / update necessary tests?
- [x] Add / update outdated documentation?

### CODE REVIEW

File: conda_rattler_solver/index.py [hunk 1]
```diff
@@ -85,6 +86,9 @@ def __init__(
             self._index.update(
                 {info.noauth_url: info for info in self._load_pkgs_cache(pkgs_dirs)}
             )
+        if installed_records:
+            installed_records = self._load_installed_records(installed_records)
+            self._index.update({info.noauth_url: info for info in installed_records})
```
Comment (kenodegard) [inline conda_rattler_solver/index.py]: The input and return are incompatible types:
- `installed_records: Iterable[PackageRecord]`
- `self._load_installed_records(...): list[_ChannelRepoInfo]`

```suggestion
            repo_infos = self._load_pkgs_cache(pkgs_dirs)
            self._index.update({info.noauth_url: info for info in repo_infos})
        if installed_records:
            repo_infos = self._load_installed_records(installed_records)
            self._index.update({info.noauth_url: info for info in repo_infos})
```

File: conda_rattler_solver/index.py [hunk 2]
```diff
@@ -82,9 +83,11 @@ def __init__(
...
         self._index.update(self._load_channels())
         if pkgs_dirs:
-            self._index.update(
-                {info.noauth_url: info for info in self._load_pkgs_cache(pkgs_dirs)}
-            )
+            repo_infos = self._load_pkgs_cache(pkgs_dirs)
+            self._index.update({info.noauth_url: info for info in repo_infos})
+        if installed_records:
+            repo_infos = self._load_installed_records(installed_records)
+            self._index.update({f"installed:{info.noauth_url}": info for info in repo_infos})
```
Comment (soapy1) [inline conda_rattler_solver/index.py]: Adding this `installed:` prefix means that if a channel is already loaded it won't get clobbered by this dict update. For example, without this prefix you may observe that the solver is missing some packages in the solve. 

This is pretty clunky, I'm not a fan. Some alternatives I've tried (that also do not spark joy):
* changing the type of `self._index` to be `list[_ChannelRepoInfo]`
* key on a different piece of information - maybe `repo_infos.local_json`. 
* ensure uniquness by suffixing the key with a number. For example, `f{info.noauthurl}_1`

Will consider some other approaches here as well.

File: conda_rattler_solver/index.py [hunk 3]
```diff
@@ -120,23 +123,15 @@ def n_packages(
...
                 for record in info.repo.load_all_records(self._package_format):
                     if filter_(record):
                         count += 1
         else:
-            for info in repos or self._index.values():
+            for info in repos or self._index:
                 count += info.repo.record_count(self._package_format)
         return count
 
-    def get_info(self, key: str) -> _ChannelRepoInfo:
```
Comment (soapy1) [inline conda_rattler_solver/index.py:132]: This is not used anywhere.

File: tests/test_solver.py [hunk 4]
```diff
@@ -856,3 +856,24 @@ def test_python_does_not_change_unless_wanted(
         link_names = {pkg["name"] for pkg in data["actions"]["LINK"]}
         unlink_names = {pkg["name"] for pkg in data["actions"]["UNLINK"]}
         assert "python" in unlink_names.intersection(link_names)
+
+
+def test_installed_packages_included_in_solver(
```
Comment (danyeaw) [inline tests/test_solver.py:861]: Hey @soapy1, your current implementation is consistent with the other tests above which is great, but do you think it is possible to use the `TmpChannelFixture` and the `mamba_repo` here to avoid network calls? This would speed up the test a ton and I don't think it would give up much for test coverage.

**Review by danyeaw — APPROVED** [review]
Hey @soapy1, thanks, this looks great! Just a minor / non-blocking comment about network calls used by one of the new tests.

**Review by jaimergp — CHANGES_REQUESTED** [review]
I need to take a deeper look but I don't think we need this. The rattler APIs are different than libmamba's. Please don't merge until I review next week.

**Comment by soapy1** [comment]
@jaimergp thanks for reviewing this! I'm definitely not super confident that this is the right approach. Sharing some notes:

I was looking at some alternative ways at passing information related to what is already installed in the conda prefix to the rattler solver.
* currently, c-r-s uses [`rattler.solve_with_sparse_repodata`](https://github.com/conda/rattler/blob/main/py-rattler/rattler/solver/solver.py#L143). It allows callers to specify the set of available packages with the [`sparse_repodata`](https://github.com/conda/rattler/blob/main/py-rattler/rattler/repo_data/sparse.py#L48) parameter
* py-rattler also has a [`solve`](https://github.com/conda/rattler/blob/main/py-rattler/rattler/solver/solver.py#L24) function which exposes a `sources` parameter to allow a caller to specify sources to query for a packages. It can take a few types, including [`Channel`'s](https://github.com/conda/rattler/blob/main/py-rattler/rattler/channel/channel.py#L11) and [`RepoDataSource`'s](https://github.com/conda/rattler/blob/main/py-rattler/rattler/repo_data/source.py#L14).

digging more into py-rattler `rattler.solve`:
* the `RepoDataSource` is pretty compelling. Like, could represent the already installed packages as a repodata source and pass it in alongside some channel sources.
* this would mean giving up using `solve_with_sparse_repodata`. Right now, c-r-s uses the `sparse_repodata` option to make [sharded repodata](https://github.com/conda/conda-rattler-solver/blob/main/conda_rattler_solver/index.py#L278) and [loading packages from the package cache](https://github.com/conda/conda-rattler-solver/blob/main/conda_rattler_solver/index.py#L307) work.
* I think moving away from `solve_with_spare_repodata` towards `solve` can work with c-r-s. `rattler.solve` also exposes the ability to set the [gateway](https://github.com/conda/rattler/blob/main/py-rattler/rattler/solver/solver.py#L27) used for getting repodata. There are gateways that [support sharded repodata](https://github.com/conda/rattler/blob/28f6d3e00d22b37339f76359201bb20b219e5207/py-rattler/rattler/repo_data/gateway.py#L87)
* this will also avoid the current c-r-s sharded implementation that [write a bunch of temporary files for the repodata that is recomposed from shards](https://github.com/conda/conda-rattler-solver/blob/main/conda_rattler_solver/index.py#L341).

**Comment by baszalmstra** [comment]
@soapy1 I actually tried this, but the overhead of marshaling the data from `msgpack.zst` -> Python -> Rust is significant. I explicitly introduced `RepoDataSource`s for this usecase but I have not been able to make this faster than the current approach with the temporary files. 

One thing we could improve, though, is to allow the `rattler.solve` function to additionally use an `SparseRepodata` object directly as an additional argument. We can then unify all these different methods under a single API. 

For what it's worth, at least on the Rust side, `SparseRepodata` can also work with in-memory data, which would at least forgo temporary files on disk.

Even better would be to let `py-rattler` handle the orchestration of determining which shards should be downloaded and the parsing of the shards. But I can see how that would not be ideal at this point.

**Comment by soapy1** [comment]
>Even better would be to let py-rattler handle the orchestration of determining which shards should be downloaded and the parsing of the shards. But I can see how that would not be ideal at this point.

ah, I see. I was under the impression that gateway already had this functionality.

> One thing we could improve, though, is to allow the rattler.solve function to additionally use an SparseRepodata object directly as an additional argument. We can then unify all these different methods under a single API.

I really like this idea, trying it out!

**Comment by soapy1** [comment]
Trying out using the `RepoDataSource`'s api in this branch https://github.com/conda/conda-rattler-solver/compare/main...soapy1:add-installed-records-to-solve-pool-2?expand=1. Depends on changes in rattler from https://github.com/conda/rattler/compare/main...soapy1:rattler:solve-sparse-repodata?expand=1.

This is just a rough demo, I think there is a more elegant way to structure this change in c-r-s. I did some initial benchmarking and it looks like this is not introducing new performance issues. 

edit: below are some results from a quick test (ref: https://github.com/conda/conda-rattler-solver/pull/42#issuecomment-3828875795)

### this pr

```
$ hyperfine -w 2 "python -m conda create -d jupyterlab -c conda-forge --override-channels --solver=rattler"
Benchmark 1: python -m conda create -d jupyterlab -c conda-forge --override-channels --solver=rattler
  Time (mean ± σ):      2.052 s ±  0.051 s    [User: 1.796 s, System: 0.173 s]
  Range (min … max):    1.999 s …  2.150 s    10 runs
```

```
$ hyperfine -w 2 "python -m conda create -d vaex -c conda-forge --override-channels --solver=rattler"      

Benchmark 1: python -m conda create -d vaex -c conda-forge --override-channels --solver=rattler
  Time (mean ± σ):     10.126 s ±  0.225 s    [User: 9.391 s, System: 0.653 s]
  Range (min … max):    9.776 s … 10.595 s    10 runs
```

### using the alternate demo
ref: https://github.com/conda/rattler/pull/2627 and https://github.com/conda/conda-rattler-solver/compare/main...soapy1:add-installed-records-to-solve-pool-2?expand=1

```
$ hyperfine -w 2 "python -m conda create -d jupyterlab -c conda-forge --override-channels --solver=rattler"
Benchmark 1: python -m conda create -d jupyterlab -c conda-forge --override-channels --solver=rattler
  Time (mean ± σ):      1.925 s ±  0.025 s    [User: 2.005 s, System: 0.233 s]
  Range (min … max):    1.890 s …  1.958 s    10 runs
 ```

```
$ hyperfine -w 2 "python -m conda create -d vaex -c conda-forge --override-channels --solver=rattler"
Benchmark 1: python -m conda create -d vaex -c conda-forge --override-channels --solver=rattler
  Time (mean ± σ):      8.735 s ±  0.129 s    [User: 9.270 s, System: 0.727 s]
  Range (min … max):    8.453 s …  8.925 s    10 runs
```

### on main

```
 $ hyperfine -w 2 "python -m conda create -d jupyterlab -c conda-forge --override-channels --solver=rattler"

Benchmark 1: python -m conda create -d jupyterlab -c conda-forge --override-channels --solver=rattler
  Time (mean ± σ):      2.077 s ±  0.051 s    [User: 1.797 s, System: 0.178 s]
  Range (min … max):    2.007 s …  2.151 s    10 runs
```

```
$ hyperfine -w 2 "python -m conda create -d vaex -c conda-forge --override-channels --solver=rattler"

Benchmark 1: python -m conda create -d vaex -c conda-forge --override-channels --solver=rattler
  Time (mean ± σ):      9.930 s ±  0.173 s    [User: 9.203 s, System: 0.629 s]
  Range (min … max):    9.764 s … 10.352 s    10 runs
```

File: conda_rattler_solver/index.py [hunk 5]
```diff
@@ -357,9 +351,71 @@ def _load_pkgs_cache(self, pkgs_dirs: PathsType) -> list[_ChannelRepoInfo]:
...
+        Returns the list of _ChannelRepoInfo object that contains a rattler.SparseRepoData
+        object that can be used to query the installed packages.
+        """
+        repos = []
+        records_map = {}
+        for record in installed_records:
+            if record.subdir not in self._subdirs:
+                continue
+            record_data = dict(record.dump())
+            for field in (
```
Comment (ryanskeith) [inline conda_rattler_solver/index.py:373]: Just a small optimization.  It looks like that [platform](https://github.com/conda/conda/blob/db924d4079101deb2d45dd60a39ae0b1979864c9/conda/models/records.py#L407), [size](https://github.com/conda/conda/blob/db924d4079101deb2d45dd60a39ae0b1979864c9/conda/models/records.py#L460) and [timestamp](https://github.com/conda/conda/blob/db924d4079101deb2d45dd60a39ae0b1979864c9/conda/models/records.py#L438) should always be included in the dump so it could be omitted here.

**Review by ryanskeith — APPROVED** [review]
Overall looks good to me.  I have a small suggestion for optimization.

File: conda_rattler_solver/solver.py [hunk 6]
```diff
@@ -239,11 +239,13 @@ def _collect_all_metadata(
         subdirs: Iterable[str],
         in_state: SolverInputState,
     ) -> RattlerIndexHelper:
+
         index = RattlerIndexHelper(
             channels=[*conda_build_channels, *channels],
             subdirs=subdirs,
             repodata_fn=self._repodata_fn,
             pkgs_dirs=context.pkgs_dirs if context.offline else (),
+            installed_records=(*in_state.installed.values(),),
```
Comment (jezdez) [inline conda_rattler_solver/solver.py:248]: Passing every installed record into the ordinary repodata pool does not give these records the fallback semantics we need. I reproduced two failures on this head:

- With strict channel priority, an active channel containing `foo=1` excludes installed `foo=2` from its removed origin channel, even when installed `bar` requires `foo>=2`.
- With `CONDA_USE_ONLY_TAR_BZ2=true`, installed `.conda` records are filtered out of this pool even though they require no download.

Could we keep the initial solve unchanged, carry a separate `missing_installed: set[str]` across retries, add a name to it only in the existing `No candidates were found` branch when the spec matches the installed record, and add only those records to `locked_packages` in `_collect_specs_main`? Rattler inserts locked records directly into the candidate pool, so that path succeeds under both settings without locking every package during a normal `update --all` solve. Keeping this separate from `out_state.conflicts` also preserves the existing no-history transitive-drop behavior.

File: tests/test_index.py [hunk 7]
```diff
@@ -116,6 +117,86 @@ def test_reload_channels(tmp_path: Path):
     assert index.n_packages() == initial_count + 1
 
 
+def _installed_record(name: str, channel_url: str, subdir: str) -> PackageRecord:
+    return PackageRecord(
+        name=name,
+        version="1.0",
+        build="0",
+        build_number=0,
+        channel=f"{channel_url}/{subdir}",
```
Comment (jezdez) [inline tests/test_index.py]: These test records embed the subdir in `Channel`, but real PrefixRecords commonly keep `channel` at the channel root and expose the subdir only as `record.subdir`. That masks the grouping bug below. Could we make the fixture use the real shape?

```suggestion
        channel=channel_url,
```

**Review by jezdez — COMMENTED** [review]
Thanks for working through this, Sophia. The issue is real, and the default reproduction now succeeds. I found two supported solver configurations where installed records still disappear as usable candidates. Both come from representing prefix records as ordinary channel repodata.

With strict channel priority, a package in the active channel excludes a compatible installed dependency from its removed origin channel. With `CONDA_USE_ONLY_TAR_BZ2=true`, an installed `.conda` record is filtered out even though keeping it requires no download.

I think the fix should keep the first solve unchanged, then add only an installed record named by a `No candidates were found` failure to `locked_packages` on the retry. Rattler inserts locked records directly into the candidate pool. This avoids channel priority and package format filtering while preserving normal `update --all` behavior for packages that do have current candidates. The missing-installed condition should remain separate from generic conflicts so we do not lock every relaxed conflict.

I left direct suggestions for the concrete channel and subdir bug in the current implementation, plus comments for the solver semantics and deterministic coverage. Those suggestions fix the metadata bug but not the central fallback semantics. The `repoquery` relaxation is unrelated and should move separately. I would keep this draft until the fallback is represented outside ordinary repodata.

File: tests/test_index.py [hunk 8]
```diff
@@ -116,6 +117,86 @@ def test_reload_channels(tmp_path: Path):
...
+    pkgs = [pkg for pkg in index.search("bar")]
+    assert len(pkgs) == 1
+    assert pkgs[0].version == "1.0"
+
+
+def test_installed_records_grouped_by_channel_and_subdir():
+    installed = (
+        _installed_record("foo", "https://conda.anaconda.org/chan-a", "linux-64"),
+        _installed_record("baz", "https://conda.anaconda.org/chan-a", "linux-64"),
+        _installed_record("bar", "https://conda.anaconda.org/chan-b", "noarch"),
```
Comment (jezdez) [inline tests/test_index.py]: Please put the noarch record under the same channel as the native records so this actually verifies grouping by `(channel, subdir)`. This fails with the current implementation once the fixture uses a root channel.

```suggestion
        _installed_record("bar", "https://conda.anaconda.org/chan-a", "noarch"),
```

File: tests/test_index.py [hunk 9]
```diff
@@ -116,6 +117,86 @@ def test_reload_channels(tmp_path: Path):
...
+    )
+    index = RattlerIndexHelper(
+        channels=(), subdirs=("linux-64", "noarch"), installed_records=installed
+    )
+    # foo and baz share a (channel, subdir) pair and must collapse into a single repo entry
+    assert len(index._index) == 2
+    assert index.n_packages() == 3
```
Comment (jezdez) [inline tests/test_index.py:176]: The package count does not catch URL rewriting because record equality ignores the URL. These assertions cover the native-to-noarch corruption directly.

```suggestion
    assert index.n_packages() == 3
    assert next(index.search("foo")).url.endswith("/linux-64/foo-1.0-0.tar.bz2")
    assert next(index.search("bar")).url.endswith("/noarch/bar-1.0-0.tar.bz2")
```

File: conda_rattler_solver/index.py [hunk 10]
```diff
@@ -357,9 +351,71 @@ def _load_pkgs_cache(self, pkgs_dirs: PathsType) -> list[_ChannelRepoInfo]:
...
+                records_map[record.channel] = empty_repodata_dict(
+                    record.subdir
+                )  # , base_url=record.channel.canonical_name)
+            records_map[record.channel][packages_key][record.fn] = record_data
+
+        for channel, repodata in records_map.items():
+            with NamedTemporaryFile(suffix=".json", delete=False, mode="w") as f:
+                f.write(json_dump(repodata))
+            subdir = channel.subdir or "noarch"
+            noauth_url = channel.urls(with_credentials=False, subdirs=(subdir,))[0]
```
Comment (jezdez) [inline conda_rattler_solver/index.py]: With the real PrefixRecord shape, keying only by channel collapses native and noarch records. `channel.subdir or "noarch"` then emits native records through a noarch source and can manufacture an invalid download URL. Keying by both values and using the record subdir fixes that concrete bug. I applied this exact replacement locally with the realistic fixture and the focused tests pass.

```suggestion
            channel_and_subdir = (record.channel, record.subdir)
            if channel_and_subdir not in records_map:
                records_map[channel_and_subdir] = empty_repodata_dict(record.subdir)
            records_map[channel_and_subdir][packages_key][record.fn] = record_data

        for (channel, subdir), repodata in records_map.items():
            with NamedTemporaryFile(suffix=".json", delete=False, mode="w") as f:
                f.write(json_dump(repodata))
            noauth_url = channel.urls(with_credentials=False, subdirs=(subdir,))[0]
```

This only fixes the subdir reconstruction. Special provenance such as `<unknown>` can still be rewritten through `channel.urls()`, which is another reason I think installed fallback records should not be modeled as ordinary channel repodata.

File: tests/test_solver.py [hunk 11]
```diff
@@ -856,3 +856,24 @@ def test_python_does_not_change_unless_wanted(
...
+
+    ref: https://github.com/conda/conda-rattler-solver/issues/88
+    """
+    with tmp_env("imagesize", "--override-channels", "--channel=defaults") as prefix:
+        _, err, rc = conda_cli(
+            "update",
+            "--all",
+            f"--prefix={prefix}",
+            "--override-channels",
+            "--channel=conda-canary",
```
Comment (jezdez) [inline tests/test_solver.py]: This depends on live `defaults` and `conda-canary`, then only checks the return code and stderr. Channel contents can change and make the test pass without exercising the fallback. Could we use local repodata and assert that the installed record and its URL remain unchanged? A deterministic version can create `test-package` from `tests/data/mamba_repo`, switch to a local empty channel, and assert the no-op result. Please also cover the two cases that currently fail on this head: strict priority with an incompatible same-name candidate in the active channel, and `CONDA_USE_ONLY_TAR_BZ2=true` with an installed `.conda` record.

File: tests/test_repoquery.py [hunk 12]
```diff
@@ -21,7 +21,16 @@ def test_repoquery():
...
+    python = [p for p in data["result"]["pkgs"] if p["name"] == "python"]
+    # Presence is required; uniqueness is not.
+    # Duplicate `python` entries can appear when the queried package (e.g. recent
+    # conda-forge builds of conda) depends on both unconstrained `python` and a
+    # pinned `python X.Y.* *_cpython`. `repoquery depends` resolves each MatchSpec
+    # independently, so both (e.g. 3.14 and 3.10) may show up.
+    # Upstream: https://github.com/mamba-org/mamba/issues/4346
+    # Same carve-out pattern in conda-libmamba-solver tests/test_repoquery.py
+    # (https://github.com/conda/conda-libmamba-solver/pull/964).
+    assert python
```
Comment (jezdez) [inline tests/test_repoquery.py:33]: This network-sensitive `repoquery` relaxation is unrelated to making installed records available during a solve. Could we move it to a separate PR so this change keeps a focused failure surface?

File: conda_rattler_solver/solver.py (hunk 6 above)
Comment (jaimergp) [reply conda_rattler_solver/solver.py:248]: The installed channel can be added with the lowest priority to minimise this problem, right? If candidates are available in higher priority in strict mode, they will be replaced (if needed). If they are not available, the low prio installed kicks in and leave things as they were.

Comment (soapy1) [reply conda_rattler_solver/solver.py:248]: @jezdez @jaimergp can you take a look at these 2 new tests I added [test_channel_priority_keeps_installed_dependency_from_removed_channel](https://github.com/conda/conda-rattler-solver/pull/101/changes#diff-728401eafc8c8e4cf40b36fc6c90ddfa9583fab597a3e2ab6b3c5c3d8f085eeeR891) and [test_channel_priority_updates_installed_dependency](https://github.com/conda/conda-rattler-solver/pull/101/changes#diff-728401eafc8c8e4cf40b36fc6c90ddfa9583fab597a3e2ab6b3c5c3d8f085eeeR964). I think they describe the intended solver behaviour, including the strict channel priority/dep update issue that Jannis points out. 
These tests should be passing on this branch now, but could use another set of eyes to make sure I'm not capturing the issue incorrectly.

Comment (soapy1) [reply conda_rattler_solver/solver.py:248]: Added a new test here that recreates one of these edge cases. [`test_explicit_update_keeps_installed_package_whose_channel_is_gone`](https://github.com/conda/conda-rattler-solver/pull/101/changes#diff-728401eafc8c8e4cf40b36fc6c90ddfa9583fab597a3e2ab6b3c5c3d8f085eeeR1046). 

This fails with the error message
```
conda_rattler_solver.exceptions.RattlerUnsatisfiableError: Cannot solve the request because of: The following packages are incompatible
└─ bar * cannot be installed because there are no viable options:
   └─ bar 1.0 would require
      └─ foo >=2, which cannot be installed because there are no viable options:
         └─ foo 2.0 is excluded because due to strict channel priority not using this option from: 'file:///tmp/pytest-of-sophia/pytest-1/test_explicit_update_keeps_ins0/channel-b/'
```

`foo 2.0` is available in the index in the lowest priority channel (aka, already installed "channel").  But is blocked from being added because of the `foo 1.0` in channel_a which is shadowing it.  `bar` is pulled in from the already installed set of package because it doesn't exist on `channel-a`. It's dependency on `foo>=2` makes the situation unsatisfiable.
